<a href="https://colab.research.google.com/github/VjorgeSN/Tensorflow/blob/main/ModeloExemplo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Instalação do tf model maker para criar o modelo
!pip install tflite-model-maker

In [4]:
#Importação dos pacotes necessários
import numpy as np

import tensorflow as tf
assert tf.__version__.startswith('2')

from tflite_model_maker import configs
from tflite_model_maker import ExportFormat
from tflite_model_maker import image_classifier
from tflite_model_maker import ImageClassifierDataLoader
from tflite_model_maker import model_spec

import matplotlib.pyplot as plt

In [5]:
#Vamos pegar algumas imagens para brincar com este exemplo simples de ponta a ponta. 
#Centenas de imagens é um bom começo para Model Maker, enquanto mais dados podem alcançar melhor precisão.

image_path = tf.keras.utils.get_file(
    'flower_photos',
    'https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz',
    untar=True
)

228818944/228813984 [==============================] - 1s 0us/step


In [6]:
#Etapa 1. Carregar dados de entrada específicos para um aplicativo ML no dispositivo. 
#Dividindo-o em dados de treinamento e dados de teste.

data = ImageClassifierDataLoader.from_folder(image_path)
train_data, test_data = data.split(0.9)

INFO:tensorflow:Load image with size: 3670, num_label: 5, labels: daisy, dandelion, roses, sunflowers, tulips.


In [7]:
#Etapa 2. Personalize o modelo do TensorFlow.

model = image_classifier.create(train_data)

INFO:tensorflow:Retraining the models...
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
hub_keras_layer_v1v2 (HubKer (None, 1280)              3413024   
_________________________________________________________________
dropout (Dropout)            (None, 1280)              0         
_________________________________________________________________
dense (Dense)                (None, 5)                 6405      
Total params: 3,419,429
Trainable params: 6,405
Non-trainable params: 3,413,024
_________________________________________________________________
None
Epoch 1/5
103/103 [==============================] - 158s 2s/step - loss: 1.0882 - accuracy: 0.6485
Epoch 2/5
103/103 [==============================] - 157s 2s/step - loss: 0.6707 - accuracy: 0.8909
Epoch 3/5
103/103 [==============================] - 157s 2s/step - loss: 0.6239 - accuracy: 0.9165
Epoch 4/5
103/103 [=======

In [8]:
#Etapa 3. Avalie o modelo.

loss, accuracy = model.evaluate(test_data)

12/12 [==============================] - 21s 1s/step - loss: 0.6066 - accuracy: 0.9155


In [9]:
#Etapa 4. Exportar para o modelo TensorFlow Lite.
#Aqui, exportamos o modelo TensorFlow Lite com metadados que fornecem um padrão para as descrições do modelo. O arquivo de etiqueta está embutido em metadados.
#Você pode baixá-lo na barra lateral esquerda da mesma forma que a parte de upload para seu próprio uso.

model.export(export_dir='.')

Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.


Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.


Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.


Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.


INFO:tensorflow:Assets written to: /tmp/tmpgzrh03k9/assets


INFO:tensorflow:Assets written to: /tmp/tmpgzrh03k9/assets


INFO:tensorflow:Label file is inside the TFLite model with metadata.


INFO:tensorflow:Label file is inside the TFLite model with metadata.


INFO:tensorflow:Saving labels in /tmp/tmp7fpj39f8/labels.txt.


INFO:tensorflow:Saving labels in /tmp/tmp7fpj39f8/labels.txt.


#Processo detalhado

Etapa 1: carregar dados de entrada específicos para um aplicativo de ML no dispositivo
O conjunto de dados de flores contém 3670 imagens pertencentes a 5 classes. Baixe a versão do arquivo do conjunto de dados e descompacte-o.

O conjunto de dados tem a seguinte estrutura de diretório:



<pre>
<b>flower_photos</b>
|__ <b>daisy</b>
    |______ 100080576_f52e8ee070_n.jpg
    |______ 14167534527_781ceb1b7a_n.jpg
    |______ ...
|__ <b>dandelion</b>
    |______ 10043234166_e6dd915111_n.jpg
    |______ 1426682852_e62169221f_m.jpg
    |______ ...
|__ <b>roses</b>
    |______ 102501987_3cdb8e5394_n.jpg
    |______ 14982802401_a3dfb22afb.jpg
    |______ ...
|__ <b>sunflowers</b>
    |______ 12471791574_bb1be83df4.jpg
    |______ 15122112402_cafa41934f.jpg
    |______ ...
|__ <b>tulips</b>
    |______ 13976522214_ccec508fe7.jpg
    |______ 14487943607_651e8062a1_m.jpg
    |______ ...
</pre>

In [ ]:
#image_path = tf.keras.utils.get_file(
      #'flower_photos',
      #'https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz',
      #untar=True)

Use a classe ImageClassifierDataLoader para carregar dados.

Quanto ao método from_folder (), ele pode carregar dados da pasta. Ele assume que os dados de imagem da mesma classe estão no mesmo subdiretório e o nome da subpasta é o nome da classe. Atualmente, imagens codificadas em JPEG e imagens codificadas em PNG são suportadas.

In [ ]:
#data = ImageClassifierDataLoader.from_folder(image_path)

Divida em dados de treinamento (80%), dados de validação (10%, opcional) e dados de teste (10%).

In [ ]:
#train_data, rest_data = data.split(0.8)
#validation_data, test_data = rest_data.split(0.5)

Mostra 25 exemplos de imagem com rótulos.

In [ ]:
#plt.figure(figsize=(10,10))
#for i, (image, label) in enumerate(data.dataset.take(25)):
  #plt.subplot(5,5,i+1)
  #plt.xticks([])
  #plt.yticks([])
  #plt.grid(False)
  #plt.imshow(image.numpy(), cmap=plt.cm.gray)
  #plt.xlabel(data.index_to_label[label.numpy()])
#plt.show()

Etapa 2: personalizar o modelo do TensorFlow
Crie um modelo de classificador de imagem personalizado com base nos dados carregados. O modelo padrão é EfficientNet-Lite0.

In [ ]:
#model = image_classifier.create(train_data, validation_data=validation_data)

Vendo a estrutura detalhada do modelo.

In [ ]:
#model.summary()

Etapa 3: Avalie o modelo personalizado
Avalie o resultado do modelo, obtenha a perda e a precisão do modelo.

In [ ]:
#loss, accuracy = model.evaluate(test_data)

Podemos traçar os resultados previstos em 100 imagens de teste. Rótulos previstos com cor vermelha são os resultados previstos incorretos, enquanto outros estão corretos.

In [ ]:
# Uma função auxiliar que retorna 'vermelho' / 'preto' dependendo se suas duas entradas
# parâmetro corresponde ou não.
#def get_label_color(val1, val2):
  #if val1 == val2:
    #return 'black'
  #else:
    #return 'red'

# Em seguida, plote 100 imagens de teste e seus rótulos previstos.
# Se um resultado de previsão for diferente do rótulo fornecido, rótulo em "teste"
# conjunto de dados, iremos destacá-lo na cor vermelha.
#plt.figure(figsize=(20, 20))
#predicts = model.predict_top_k(test_data)
#for i, (image, label) in enumerate(test_data.dataset.take(100)):
  #ax = plt.subplot(10, 10, i+1)
  #plt.xticks([])
  #plt.yticks([])
  #plt.grid(False)
  #plt.imshow(image.numpy(), cmap=plt.cm.gray)

  #predict_label = predicts[i][0][0]
  #color = get_label_color(predict_label,
                          #test_data.index_to_label[label.numpy()])
  #ax.xaxis.label.set_color(color)
  #plt.xlabel('Predicted: %s' % predict_label)
#plt.show()

Se a precisão não atender aos requisitos do aplicativo, pode-se consultar o Uso avançado para explorar alternativas, como mudar para um modelo maior, ajustar os parâmetros de retreinamento etc.

Link uso avançado: https://colab.research.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/g3doc/tutorials/model_maker_image_classification.ipynb?hl=pt_br#scrollTo=zNDBP2qA54aK

Etapa 4: exportar para o modelo TensorFlow Lite
Converta o modelo existente para o formato de modelo TensorFlow Lite com metadados. O nome de arquivo TFLite padrão é model.tflite.

In [ ]:
#model.export(export_dir='.')

See example applications and guides of image classification for more details about how to integrate the TensorFlow Lite model into mobile apps.
[example applications and guides of image classification](https://www.tensorflow.org/lite/models/image_classification/overview#example_applications_and_guides)

Os formatos de exportação permitidos podem ser um ou uma lista dos seguintes:

ExportFormat.TFLITE
ExportFormat.LABEL
ExportFormat.SAVED_MODEL
Por padrão, ele apenas exporta o modelo TensorFlow Lite com metadados. Você também pode exportar arquivos diferentes seletivamente. Por exemplo, exportando apenas o arquivo de etiqueta da seguinte forma:

In [ ]:
#model.export(export_dir='.', export_format=ExportFormat.LABEL)

Você também pode avaliar o modelo tflite com o método evaluate_tflite.

In [ ]:
#model.evaluate_tflite('model.tflite', test_data)

## Uso Avançado

A função `create` é a parte crítica desta biblioteca. Ele usa a aprendizagem por transferência com um modelo pré-treinado semelhante ao [tutorial] (https://www.tensorflow.org/tutorials/images/transfer_learning).

A função `create` contém as seguintes etapas:

1. Divida os dados em treinamento, validação e teste de dados de acordo com o parâmetro `validation_ratio` e` test_ratio`. Os valores padrão de `validation_ratio` e` test_ratio` são `0.1` e` 0.1`.
2. Faça download de um [Vetor de recurso de imagem] (https://www.tensorflow.org/hub/common_signatures/images#image_feature_vector) como o modelo básico do TensorFlow Hub. O modelo pré-treinado padrão é EfficientNet-Lite0.
3. Adicione um classificador head com Dropout Layer com `dropout_rate` entre a camada head e o modelo pré-treinado. O `dropout_rate` padrão é o valor` dropout_rate` padrão de [make_image_classifier_lib] (https://github.com/tensorflow/hub/blob/master/tensorflow_hub/tools/make_image_classifier/make_image_classifier_lib.py#L55) por TensorFlow Hub.
4. Pré-processe os dados de entrada brutos. Atualmente, as etapas de pré-processamento incluem normalizar o valor de cada pixel da imagem para modelar a escala de entrada e redimensioná-la para modelar o tamanho de entrada. EfficientNet-Lite0 tem a escala de entrada `[0, 1]` e o tamanho da imagem de entrada `[224, 224, 3]`.
5. Alimente os dados no modelo do classificador. Por padrão, os parâmetros de treinamento, como épocas de treinamento, tamanho do lote, taxa de aprendizagem, momento são os valores padrão de [make_image_classifier_lib] (https://github.com/tensorflow/hub/blob/master/tensorflow_hub/tools/make_image_classifier/make_image_classifier_lib .py # L55) do TensorFlow Hub. Apenas o cabeçote do classificador é treinado.


Nesta seção, descrevemos vários tópicos avançados, incluindo a mudança para um modelo de classificação de imagem diferente, alterando os hiperparâmetros de treinamento etc.

#Quantização pós-treinamento no modelo TensorFLow Lite

[Quantização pós-treinamento] (https://www.tensorflow.org/lite/performance/post_training_quantization) é uma técnica de conversão que pode reduzir o tamanho do modelo e a latência de inferência, ao mesmo tempo que melhora a latência do acelerador de hardware e CPU, com pouca degradação no modelo precisão. Portanto, é amplamente utilizado para otimizar o modelo.

O Model Maker oferece suporte a várias opções de quantização pós-treinamento. Vamos tomar a quantização inteira inteira como uma instância. Primeiro, defina a configuração de quantização para forçar a quantização inteira completa para todas as operações, incluindo a entrada e a saída. O tipo de entrada e o tipo de saída são `uint8` por padrão. Você também pode alterá-los para outros tipos como `int8` definindo` inference_input_type` e `inference_output_type` na configuração.

In [ ]:
#config = configs.QuantizationConfig.create_full_integer_quantization(representative_data=test_data, is_integer_only=True)

Em seguida, exportamos o modelo TensorFlow Lite com essa configuração.

In [ ]:
#model.export(export_dir='.', tflite_filename='model_quant.tflite', quantization_config=config)

No Colab, você pode baixar o modelo chamado model_quant.tflite da barra lateral esquerda, igual à parte de upload mencionada acima.

#Mudar o modelo

Mude para o modelo compatível com esta biblioteca.
Esta biblioteca é compatível com os modelos EfficientNet-Lite, MobileNetV2, ResNet50 agora. EfficientNet-Lite é uma família de modelos de classificação de imagem que podem alcançar precisão de última geração e adequados para dispositivos Edge. O modelo padrão é EfficientNet-Lite0.

Poderíamos mudar o modelo para MobileNetV2 apenas definindo o parâmetro model_spec para mobilenet_v2_spec no método de criação.

In [ ]:
#model = image_classifier.create(train_data, model_spec=model_spec.mobilenet_v2_spec, validation_data=validation_data)

Avalie o modelo MobileNetV2 recém-treinado para ver a precisão e a perda nos dados de teste.

In [ ]:
#loss, accuracy = model.evaluate(test_data)

#Mudança para o modelo no TensorFlow Hub

Além disso, também poderíamos mudar para outros novos modelos que inserem uma imagem e geram um vetor de recursos com o formato TensorFlow Hub.

Como modelo [Inception V3] (https://tfhub.dev/google/imagenet/inception_v3/feature_vector/1) como um exemplo, poderíamos definir `inception_v3_spec` que é um objeto de` ImageModelSpec` e contém a especificação do Inception Modelo V3.

Precisamos especificar o nome do modelo `name`, o url do modelo TensorFlow Hub` uri`. Enquanto isso, o valor padrão de `input_image_shape` é` [224, 224] `. Precisamos alterá-lo para `[299, 299]` para o modelo Inception V3.

In [ ]:
#inception_v3_spec = model_spec.ImageModelSpec(
    #uri='https://tfhub.dev/google/imagenet/inception_v3/feature_vector/1')
#inception_v3_spec.input_image_shape = [299, 299]

Então, definindo o parâmetro `model_spec` para` inception_v3_spec` no método `create`, poderíamos treinar novamente o modelo Inception V3.

As etapas restantes são exatamente as mesmas e poderíamos obter um modelo personalizado do InceptionV3 TensorFlow Lite no final.

#Mude seu próprio modelo personalizado

Se quisermos usar o modelo personalizado que não está no TensorFlow Hub, devemos criar e exportar [ModelSpec] (https://www.tensorflow.org/hub/api_docs/python/hub/ModuleSpec) no TensorFlow Hub.

Então comece a definir o objeto `ImageModelSpec` como o processo acima.

#Alterar os hiperparâmetros de treinamento

Também podemos alterar os hiperparâmetros de treinamento como `epochs`,` dropout_rate` e `batch_size` que podem afetar a precisão do modelo. Os parâmetros do modelo que você pode ajustar são:


* `epochs`: mais épocas podem alcançar melhor precisão até que converta, mas treinar por muitas épocas pode levar a overfitting.
* `dropout_rate`: A taxa de abandono, evite overfitting. Nenhum por padrão.
* `batch_size`: número de amostras para usar em uma etapa de treinamento. Nenhum por padrão.
* `validation_data`: dados de validação. Se nenhum, ignora o processo de validação. Nenhum por padrão.
* `train_whole_model`: se verdadeiro, o módulo Hub é treinado junto com a camada de classificação no topo. Caso contrário, treine apenas a camada de classificação superior. Nenhum por padrão.
* `learning_rate`: Taxa de aprendizagem básica. Nenhum por padrão.
* `momentum`: um float do Python encaminhado para o otimizador. Só usado quando
      `use_hub_library` é True. Nenhum por padrão.
*   `shuffle`: Booleano, se os dados devem ser misturados. Falso por padrão.
*   `use_augmentation`: Booleano, usa aumento de dados para pré-processamento. Falso por padrão.
* `use_hub_library`: Booleano, use` make_image_classifier_lib` do hub tensorflow para treinar novamente o modelo. Este pipeline de treinamento pode alcançar melhor desempenho para conjuntos de dados complicados com muitas categorias. Verdadeiro por padrão.
* `warmup_steps`: Número de etapas de aquecimento para a programação de aquecimento na taxa de aprendizagem. Se Nenhum, o warmup_steps padrão é usado, que é o total de etapas de treinamento em duas épocas. Usado apenas quando `use_hub_library` é False. Nenhum por padrão.
* `model_dir`: Opcional, a localização dos arquivos de checkpoint do modelo. Usado apenas quando `use_hub_library` é False. Nenhum por padrão.

Parâmetros que são Nenhum por padrão, como `epochs`, obterão os parâmetros padrão concretos em [make_image_classifier_lib] (https://github.com/tensorflow/hub/blob/02ab9b7d3455e99e97abecf43c5d598a5528e20c/tensorflow_hub/tools/make_image_classificador de imagem54) Biblioteca do TensorFlow Hub ou [train_image_classifier_lib] (https://github.com/tensorflow/examples/blob/f0260433d133fd3cea4a920d1e53ecda07163aee/tensorflow_examples/lite/model_maker/core/task/train_imagelib.class).

Por exemplo, podemos treinar com mais épocas.

In [ ]:
#model = image_classifier.create(train_data, validation_data=validation_data, epochs=10)

Avalie o modelo recém-treinado com 10 épocas de treinamento.

In [ ]:
#loss, accuracy = model.evaluate(test_data)